### PR TITLE
feat(leaderboard): differentiate inference work vs network reward earnings

### PR DESCRIPTION
**Pull request by Gajesh2007**
## Summary

The public leaderboard (`GET /v1/leaderboard`) and network totals (`GET /v1/network/totals`) previously aggregated **only** `provider_earnings` — i.e. money earned from **inference work**. Providers also receive **non-inference rewards** for participating in the network (`referral_reward`, `admin_reward`), which live only in `ledger_entries` and never showed up on the leaderboard.

This PR makes the leaderboard surface a clear **work vs reward** breakdown plus a combined total.

New fields on both endpoints (and the store rows behind them):

| field | meaning |
|---|---|
| `earnings_micro_usd` | **total** = work + reward (now ranks the *earnings* metric) |
| `work_earnings_micro_usd` | inference payouts (`provider_earnings`) |
| `reward_earnings_micro_usd` | network rewards (`referral_reward` + `admin_reward`) |

**No double-counting:** work comes only from `provider_earnings`; rewards only from reward-type `ledger_entries` (the inference `payout` ledger entry is excluded). The earnings ranking uses the combined total with a deterministic `account_id` tiebreaker, and reward-only accounts now appear (work = 0).

## Backend (`coordinator/`)
- `store/interface.go`: `LeaderboardRow` / `NetworkTotalsRow` gain `Work`/`Reward` fields; added `RewardLedgerTypes` + `IsRewardLedgerType()` (single source of truth for which ledger types count as rewards).
- `store/postgres.go`: `Leaderboard` rewritten as a `FULL OUTER JOIN` of `provider_earnings` and a reward-ledger CTE; `NetworkTotals` unions both sources for totals and the active-account count.
- `store/memory.go`: mirrored logic.
- `api/leaderboard.go`: both handlers emit the new fields.
- `store/leaderboard_rewards_test.go` (new): work-only, reward-only, combined, ranking-by-total, non-reward exclusion, network totals, type set.

## Frontend (`console-ui/` — `/stats` Leaderboard tab)
- **Total earnings** card with a `Work $X · Rewards $Y` breakdown sub-line.
- Rankings table gains explicit **Work** and **Rewards** columns (rewards tinted amber when > 0).
- Podium cards show the split; section copy explains the two sources.
- `stats/leaderboard/format.ts` (new): pure, unit-tested helpers (`formatEarningsBreakdown`, `rewardToneClass`) + `format.test.ts`.

## Compatibility
Only consumers of these store methods are the two API handlers; only consumer of the JSON is the `/stats` page — all updated. The `earnings_micro_usd` field name is preserved (semantics widen from inference-only to total). Reward types intentionally limited to `referral_reward` + `admin_reward` (withdrawable, earnings-like); non-withdrawable `admin_credit`/`invite_credit` are excluded.

## Verification
- Backend: `gofmt` clean · `go build ./...` · `go test ./store/... ./api/...` (incl. 7 new suites, pass uncached)
- Frontend: `eslint src/` (0 errors) · vitest 305/305 (incl. 7 new) · `next build` ✓ compiled

🤖 Generated with [opencode](https://opencode.ai)

---
<a href="https://app.blacksmith.sh/Layr-Labs/codesmith/d-inference/pr/405"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1784484993&installation_id=133247490&pr_number=405&repository=Layr-Labs%2Fd-inference&return_to=https%3A%2F%2Fgithub.com%2FLayr-Labs%2Fd-inference%2Fpull%2F405&signature=d9dc892efde4763f8ddb97af1f3423fdb83300b43db4269cc90833cc0a06cb32"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

File: coordinator/store/postgres.go [hunk 1]
```diff
@@ -2068,29 +2087,50 @@ func (s *PostgresStore) Leaderboard(metric LeaderboardMetric, since time.Time, l
...
+	             COALESCE(w.work_micro,0) + COALESCE(r.reward_micro,0) AS earnings_micro_usd,
+	             COALESCE(w.work_micro,0)                              AS work_micro_usd,
+	             COALESCE(r.reward_micro,0)                            AS reward_micro_usd,
+	             COALESCE(w.tokens,0)                                  AS tokens,
+	             COALESCE(w.jobs,0)                                    AS jobs
+	      FROM work w
+	      FULL OUTER JOIN reward r ON w.account_id = r.account_id
+	      ORDER BY ` + orderCol + ` DESC, account_id ASC
 	      LIMIT $` + strconv.Itoa(len(args)+1)
 	args = append(args, limit)
```
Comment (ethenotethan) [inline coordinator/store/postgres.go:2143]: 🟡 **[MEDIUM]** ⚡ Leaderboard query lacks LIMIT validation allowing unbounded result sets

💡 _Suggestion:_ Add validation to ensure limit is within reasonable bounds (e.g., max 1000) before query execution to prevent memory exhaustion from large result sets

📊 _Score: 3×4 = 12 · Category: Unbounded allocations_

File: coordinator/store/memory.go [hunk 2]
```diff
@@ -687,51 +687,82 @@ func (s *MemoryStore) UsageTimeSeries(since time.Time) []UsageBucket {
...
+			if rows[i].EarningsMicroUSD != rows[j].EarningsMicroUSD {
+				return rows[i].EarningsMicroUSD > rows[j].EarningsMicroUSD
+			}
 		}
+		return rows[i].AccountID < rows[j].AccountID
 	})
 	if len(rows) > limit {
 		rows = rows[:limit]
 	}
 	return rows
```
Comment (ethenotethan) [inline coordinator/store/memory.go:765]: 🔵 **[INFO]** ⚡ Leaderboard implementation iterates all provider earnings and ledger entries without time-based filtering optimization

💡 _Suggestion:_ Consider adding time-based indexing or pre-filtering data structures to avoid O(n) iteration over all historical data when since parameter is specified

📊 _Score: 2×3 = 6 · Category: Inefficient data structures_

File: coordinator/store/memory.go [hunk 3]
```diff
@@ -741,16 +772,27 @@ func (s *MemoryStore) NetworkTotals(since time.Time) NetworkTotalsRow {
...
-				seen[e.AccountID] = struct{}{}
-				t.ActiveAccounts++
-			}
+			seen[e.AccountID] = struct{}{}
+		}
+	}
+	for _, e := range s.ledgerEntries {
+		if !IsRewardLedgerType(e.Type) {
+			continue
+		}
```
Comment (ethenotethan) [inline coordinator/store/memory.go:793]: 🔵 **[INFO]** ⚡ NetworkTotals implementation iterates all provider earnings and ledger entries without time-based filtering optimization

💡 _Suggestion:_ Consider adding time-based indexing or pre-filtering data structures to avoid O(n) iteration over all historical data when since parameter is specified

📊 _Score: 2×3 = 6 · Category: Inefficient data structures_

File: console-ui/src/app/stats/leaderboard/format.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+/**
+ * Tailwind text-color token for a reward value. Rewards that are actually paid
+ * out get the amber accent so they are visually differentiated from inference
+ * work; zero rewards stay muted.
+ */
+export function rewardToneClass(rewardMicroUsd: number): string {
+  return rewardMicroUsd > 0 ? "text-accent-amber" : "text-text-tertiary";
+}
```
Comment (ethenotethan) [inline console-ui/src/app/stats/leaderboard/format.ts:30]: 🔵 **[INFO]** 🧩 Separate file for two simple formatting functions that could live inline

💡 _Suggestion:_ Move these 15-line functions directly into page.tsx where they're used, or combine with existing formatters

📊 _Score: 2×3 = 6 · Category: over-abstraction_

File: coordinator/store/interface.go [hunk 5]
```diff
@@ -873,22 +873,43 @@ const (
...
+var RewardLedgerTypes = []LedgerEntryType{LedgerReferralReward, LedgerAdminReward}
+
+// IsRewardLedgerType reports whether t is counted as reward earnings.
+func IsRewardLedgerType(t LedgerEntryType) bool {
+	for _, rt := range RewardLedgerTypes {
+		if t == rt {
+			return true
+		}
+	}
+	return false
```
Comment (ethenotethan) [inline coordinator/store/interface.go:912]: 🔵 **[INFO]** 🧩 Global variable and helper function for simple type checking

💡 _Suggestion:_ Use a simple switch statement in the few places this is needed instead of maintaining a global list

📊 _Score: 2×2 = 4 · Category: over-abstraction_

File: coordinator/store/leaderboard_rewards_test.go [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+	}
+	for _, lt := range notReward {
+		if IsRewardLedgerType(lt) {
+			t.Errorf("IsRewardLedgerType(%q) = true, want false", lt)
+		}
+	}
+	if len(RewardLedgerTypes) != 2 {
+		t.Errorf("RewardLedgerTypes has %d entries, want 2", len(RewardLedgerTypes))
+	}
+}
```
Comment (ethenotethan) [inline coordinator/store/leaderboard_rewards_test.go:341]: 🔵 **[INFO]** 🧩 294-line test file with repetitive test patterns and helper functions

💡 _Suggestion:_ Use table-driven tests to reduce the 7 similar test functions to 1-2 parameterized tests

📊 _Score: 1×2 = 2 · Category: duplicate logic_

File: coordinator/store/postgres.go [hunk 7]
```diff
@@ -2057,9 +2057,28 @@ func (s *PostgresStore) UsageTimeSeries(since time.Time) []UsageBucket {
...
+func rewardLedgerTypesSQLList() string {
+	out := ""
+	for i, t := range RewardLedgerTypes {
+		if i > 0 {
+			out += ","
+		}
+		out += "'" + string(t) + "'"
+	}
+	return out
+}
```
Comment (ethenotethan) [inline coordinator/store/postgres.go:2079]: 🟡 **[MEDIUM]** 🧩 Manual SQL string building instead of using query builder or prepared fragments

💡 _Suggestion:_ Use a proper SQL builder library or at minimum extract this to a reusable query fragment constant

📊 _Score: 3×3 = 9 · Category: reinventing stdlib_

File: console-ui/src/app/stats/page.tsx [hunk 8]
```diff
@@ -2286,8 +2294,17 @@ function LeaderboardSection() {
...
+            sub={
+              totals
+                ? formatEarningsBreakdown(
+                    totals.work_earnings_micro_usd,
+                    totals.reward_earnings_micro_usd,
+                    formatUSDFromMicro,
+                  )
+                : undefined
+            }
           />
```
Comment (ethenotethan) [inline console-ui/src/app/stats/page.tsx:2308]: 🔵 **[INFO]** 🧩 Repeated formatEarningsBreakdown calls with same parameters in multiple places

💡 _Suggestion:_ Extract a helper that takes the entry/totals object and returns the formatted breakdown

📊 _Score: 2×3 = 6 · Category: duplicate logic_

**Review by ethenotethan — CHANGES_REQUESTED** [review]
## Automated Code Review — Layr-Labs/d-inference#

**Verdict:** REQUEST_CHANGES

### Security — ✅ No issues found

### Performance — 3 finding(s) (1 blocking)
- 🟡 **[MEDIUM]** `coordinator/store/postgres.go:2089-2135` — Leaderboard query lacks LIMIT validation allowing unbounded result sets
  - _Suggestion:_ Add validation to ensure limit is within reasonable bounds (e.g., max 1000) before query execution to prevent memory exhaustion from large result sets
- 🔵 **[INFO]** `coordinator/store/memory.go:687-759` — Leaderboard implementation iterates all provider earnings and ledger entries without time-based filtering optimization
  - _Suggestion:_ Consider adding time-based indexing or pre-filtering data structures to avoid O(n) iteration over all historical data when since parameter is specified
- 🔵 **[INFO]** `coordinator/store/memory.go:759-785` — NetworkTotals implementation iterates all provider earnings and ledger entries without time-based filtering optimization
  - _Suggestion:_ Consider adding time-based indexing or pre-filtering data structures to avoid O(n) iteration over all historical data when since parameter is specified

### Type_diligence — ✅ No issues found

### Additive_complexity — 5 finding(s) (1 blocking)
- 🔵 **[INFO]** `console-ui/src/app/stats/leaderboard/format.ts:1-30` — Separate file for two simple formatting functions that could live inline
  - _Suggestion:_ Move these 15-line functions directly into page.tsx where they're used, or combine with existing formatters
- 🔵 **[INFO]** `coordinator/store/interface.go:900-912` — Global variable and helper function for simple type checking
  - _Suggestion:_ Use a simple switch statement in the few places this is needed instead of maintaining a global list
- 🔵 **[INFO]** `coordinator/store/leaderboard_rewards_test.go:1-294` — 294-line test file with repetitive test patterns and helper functions
  - _Suggestion:_ Use table-driven tests to reduce the 7 similar test functions to 1-2 parameterized tests
- 🟡 **[MEDIUM]** `coordinator/store/postgres.go:2060-2073` — Manual SQL string building instead of using query builder or prepared fragments
  - _Suggestion:_ Use a proper SQL builder library or at minimum extract this to a reusable query fragment constant
- 🔵 **[INFO]** `console-ui/src/app/stats/page.tsx:2294-2308` — Repeated formatEarningsBreakdown calls with same parameters in multiple places
  - _Suggestion:_ Extract a helper that takes the entry/totals object and returns the formatted breakdown

_8 finding(s) total, 2 blocking. Verdict: REQUEST_CHANGES._

_🤖 Automated review by Centaur · DAR-186_

**Comment by Gajesh2007** [comment]
## Review feedback addressed — pushed `06f27b15`

Thanks all. Triaged every comment; summary of what changed and what I deliberately didn't.

### ✅ Fixed (both Codex P2 findings — these were the high-signal ones)

**1. Consumer/non-provider rewards polluting the provider leaderboard** (`interface.go` / Codex)
Correct and important. `referral_reward` (open to any account) and `admin_reward` could land on accounts that never provided inference, ranking non-providers on the *Provider Earnings Leaderboard* and inflating `active_accounts`. Rewards are now credited **only to providers** (accounts with inference work in the window):
- `Leaderboard`: `FULL OUTER JOIN` → `LEFT JOIN` from `provider_earnings`.
- `NetworkTotals`: reward ledger entries joined to a `providers` CTE; `active_accounts` = distinct providers.
- in-memory store mirrored; tests updated + new `TestLeaderboardExcludesNonProviderRewards`.

I verified this **read-only against prod**: it excludes **~$1,033** of reward that was being mis-attributed to non-providers, drops `active_accounts` 356 → 281, and yields **zero** leaderboard rows with `work == 0`.

**2. Missing index for reward ledger scans** (`postgres.go` / Codex)
Added a partial index:
```sql
CREATE INDEX IF NOT EXISTS idx_ledger_reward
  ON ledger_entries(account_id, created_at DESC)
  WHERE entry_type IN ('referral_reward','admin_reward');
```
Predicate is derived from `RewardLedgerTypes` so it always matches the query's IN-list. Confirmed live the ledger is ~2.1M rows with only a few hundred reward rows, so this turns a full scan into a tiny partial-index scan per cache miss.
⚠️ **Deploy note:** this is a non-concurrent build (consistent with the existing indexes + safe via the autocommit migration loop). It takes a brief `SHARE` lock on `ledger_entries` for ~1s during the one-time build. If you'd rather avoid any write stall on deploy, pre-create it `CONCURRENTLY` by hand before merge (couldn't use `CONCURRENTLY` in the migration due to the pgx extended-protocol + invalid-index-on-failure caveats).

### 🟡 Centaur "REQUEST_CHANGES" — both blocking items are false positives in context
- **"Leaderboard query lacks LIMIT validation / unbounded result sets"** — already bounded: both `store.Leaderboard` (`if limit <= 0 || limit > 200 { limit = 50 }`) and the API handler clamp to ≤200. No unbounded path exists.
- **"Use a SQL builder instead of manual string building"** — the entire coordinator uses raw pgx SQL; introducing a builder for one query would be inconsistent. The interpolated fragment is package constants (`RewardLedgerTypes`), not user input, so no injection risk.

INFO nits (kept, with rationale): the in-memory store is the dev/test fallback (prod is Postgres) and was already O(n); `RewardLedgerTypes`/`IsRewardLedgerType` is an intentional single source of truth shared by the SQL builder, the in-memory store, and the index predicate; `format.ts` is extracted specifically so it's unit-tested (AGENTS.md encourages modular helpers).

### 🔐 Threat-model bot
Endpoint is intentionally public and **pseudonymized** (pre-existing behaviour; unchanged by this PR — raw account IDs are never returned). The providers-only change above also shrinks the aggregation/privacy surface it flagged (no consumer accounts exposed via rewards). Happy to add a `docs/threat-model.yaml` entry for the public-stats surface in a follow-up if you'd like.

All green: `gofmt` clean · `go build ./...` · `go test ./store/... ./api/...` · live read-only prod verification.

File: console-ui/src/app/stats/page.tsx (hunk 8 above)
Comment (ethenotethan) [reply console-ui/src/app/stats/page.tsx:2308]: :white_check_mark: Looks addressed in `06f27b1` — The code now imports and uses formatEarningsBreakdown from a shared module instead of duplicating the logic. Resolving.

File: coordinator/store/postgres.go [hunk 9]
```diff
@@ -2068,29 +2095,50 @@ func (s *PostgresStore) Leaderboard(metric LeaderboardMetric, since time.Time, l
...
+	      )
+	      SELECT w.account_id                            AS account_id,
+	             w.work_micro + COALESCE(r.reward_micro,0) AS earnings_micro_usd,
+	             w.work_micro                            AS work_micro_usd,
+	             COALESCE(r.reward_micro,0)              AS reward_micro_usd,
+	             w.tokens                                AS tokens,
+	             w.jobs                                  AS jobs
+	      FROM work w
+	      LEFT JOIN reward r ON w.account_id = r.account_id
+	      ORDER BY ` + orderCol + ` DESC, account_id ASC
```
Comment (ethenotethan) [inline coordinator/store/postgres.go:2141]: 🟡 **[MEDIUM]** ⚡ Complex multi-CTE query without LIMIT on reward subquery could scan entire ledger table

💡 _Suggestion:_ Add LIMIT clause to reward CTE or consider using EXISTS instead of JOIN for large ledger tables

📊 _Score: 3×4 = 12 · Category: N+1 queries_

File: coordinator/store/memory.go [hunk 10]
```diff
@@ -687,70 +687,119 @@ func (s *MemoryStore) UsageTimeSeries(since time.Time) []UsageBucket {
...
+		if !since.IsZero() && e.CreatedAt.Before(since) {
+			continue
+		}
+		if row, ok := agg[e.AccountID]; ok {
+			row.RewardEarningsMicroUSD += e.AmountMicroUSD
+		}
+	}
 	rows := make([]LeaderboardRow, 0, len(agg))
```
Comment (ethenotethan) [inline coordinator/store/memory.go:740]: 🔵 **[INFO]** ⚡ Linear scan through all ledger entries for each leaderboard request

💡 _Suggestion:_ Pre-index ledger entries by type and account_id, or use a map structure for faster lookups

📊 _Score: 2×3 = 6 · Category: Inefficient data structures_

File: console-ui/src/app/stats/leaderboard/format.ts (hunk 4 above)
Comment (ethenotethan) [inline console-ui/src/app/stats/leaderboard/format.ts:30]: 🔵 **[INFO]** 🧩 Separate file for two simple formatting functions that could live inline

💡 _Suggestion:_ Move these 30 lines directly into page.tsx where they're used - the abstraction overhead (separate file, imports, dependency injection) outweighs the benefit for such simple helpers

📊 _Score: 2×3 = 6 · Category: over-abstraction_

File: coordinator/store/interface.go (hunk 5 above)
Comment (ethenotethan) [inline coordinator/store/interface.go:912]: 🔵 **[INFO]** 🧩 IsRewardLedgerType reimplements basic slice membership check

💡 _Suggestion:_ Use slices.Contains(RewardLedgerTypes, t) from Go 1.21+ stdlib instead of custom loop - removes 8 lines of boilerplate

📊 _Score: 2×2 = 4 · Category: duplicate logic_

File: coordinator/store/postgres.go [hunk 11]
```diff
@@ -2057,9 +2063,30 @@ func (s *PostgresStore) UsageTimeSeries(since time.Time) []UsageBucket {
...
+	out := ""
+	for i, t := range RewardLedgerTypes {
+		if i > 0 {
+			out += ","
+		}
+		out += "'" + string(t) + "'"
+	}
+	return out
+}
+
```
Comment (ethenotethan) [inline coordinator/store/postgres.go:2080]: 🟡 **[MEDIUM]** 🧩 Manual SQL string building instead of using query builder or prepared fragments

💡 _Suggestion:_ Use strings.Join with quoted values or a proper SQL builder - the current approach is error-prone and harder to maintain than standard approaches

📊 _Score: 3×4 = 12 · Category: reinventing stdlib_

File: coordinator/store/leaderboard_rewards_test.go [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
```
Comment (ethenotethan) [inline coordinator/store/leaderboard_rewards_test.go:341]: 🟡 **[MEDIUM]** 🧩 341-line test file with excessive test cases for relatively simple logic

💡 _Suggestion:_ Consolidate into 2-3 focused tests: basic work/reward differentiation, ranking correctness, and edge cases - current granularity creates maintenance burden without proportional value

📊 _Score: 3×3 = 9 · Category: over-configuration_

**Review by ethenotethan — CHANGES_REQUESTED** [review]
## Automated Code Review — Layr-Labs/d-inference#

**Verdict:** REQUEST_CHANGES

### Security — ✅ No issues found

### Performance — 3 finding(s) (1 blocking)
- 🟡 **[MEDIUM]** `coordinator/store/postgres.go:2095-2141` — Complex multi-CTE query without LIMIT on reward subquery could scan entire ledger table
  - _Suggestion:_ Add LIMIT clause to reward CTE or consider using EXISTS instead of JOIN for large ledger tables
- 🔵 **[INFO]** `coordinator/store/memory.go:728-740` — Linear scan through all ledger entries for each leaderboard request
  - _Suggestion:_ Pre-index ledger entries by type and account_id, or use a map structure for faster lookups
- 🔵 **[INFO]** `console-ui/src/app/stats/page.tsx:2371-2408` — formatUSDFromMicro called multiple times per row in render loop
  - _Suggestion:_ Pre-compute formatted values outside the map function or memoize the formatter

### Type_diligence — ✅ No issues found

### Additive_complexity — 4 finding(s) (2 blocking)
- 🔵 **[INFO]** `console-ui/src/app/stats/leaderboard/format.ts:1-30` — Separate file for two simple formatting functions that could live inline
  - _Suggestion:_ Move these 30 lines directly into page.tsx where they're used - the abstraction overhead (separate file, imports, dependency injection) outweighs the benefit for such simple helpers
- 🔵 **[INFO]** `coordinator/store/interface.go:895-912` — IsRewardLedgerType reimplements basic slice membership check
  - _Suggestion:_ Use slices.Contains(RewardLedgerTypes, t) from Go 1.21+ stdlib instead of custom loop - removes 8 lines of boilerplate
- 🟡 **[MEDIUM]** `coordinator/store/postgres.go:2070-2080` — Manual SQL string building instead of using query builder or prepared fragments
  - _Suggestion:_ Use strings.Join with quoted values or a proper SQL builder - the current approach is error-prone and harder to maintain than standard approaches
- 🟡 **[MEDIUM]** `coordinator/store/leaderboard_rewards_test.go:1-341` — 341-line test file with excessive test cases for relatively simple logic
  - _Suggestion:_ Consolidate into 2-3 focused tests: basic work/reward differentiation, ranking correctness, and edge cases - current granularity creates maintenance burden without proportional value

_7 finding(s) total, 3 blocking. Verdict: REQUEST_CHANGES._

_🤖 Automated review by Centaur · DAR-186_
